### PR TITLE
Add keep flag

### PR DESCRIPTION
**Pull request by timja**
**Purpose of the PR**

- Add `--keep` flag

Fixes https://github.com/Azure/acr-cli/issues/69

### CODE REVIEW

**Comment by timja** [comment]
Hi 

@Wwwsylvia 
I've just raised a support ticket 120072325000556
Which covers a bug in the API I've encountered with this.

Some images are not respecting pagination, and just return the same page, this only occurs when sortby and last are set.

This is causing an infinite loop on those images, I can put a hack in which passes in the previous last tag and if its the same as the new last tag then exit the loop.

But just wanted feedback first on what the correct approach is.

All details are in the ticket, including the registry it's occurring on and working and non working images

Thanks

**Comment by Wwwsylvia** [comment]
> Hi
> 
> @Wwwsylvia
> I've just raised a support ticket 120072325000556
> Which covers a bug in the API I've encountered with this.
> 
> Some images are not respecting pagination, and just return the same page, this only occurs when sortby and last are set.
> 
> This is causing an infinite loop on those images, I can put a hack in which passes in the previous last tag and if its the same as the new last tag then exit the loop.
> 
> But just wanted feedback first on what the correct approach is.
> 
> All details are in the ticket, including the registry it's occurring on and working and non working images
> 
> Thanks

Hi @timja, thanks for the PR and the bug report, we will check it and get back to you.

**Comment by Wwwsylvia** [comment]
Hi @timja, not sure how you set the `last` parameter, did you grab the url from the returned `Link` header? 

For example, in below case, the last tag of the first page is `12.04`.
To get the next page, we should query `/acr/v1/library/ubuntu/_tags?last=8586163895140836946%2612.04&n=3&orderby=timedesc`, which is encoded in the `Link` header. (Notice that the `last` parameter here is `8586163895140836946%2612.04`, not `12.04`.)

``` bash
$ curl -s -u ***:*** 'https://myregistry.azurecr.io/acr/v1/library/ubuntu/_tags?orderby=timedesc&last=&n=3' -v | jq .
...
< Link: </acr/v1/library/ubuntu/_tags?last=8586163895140836946%2612.04&n=3&orderby=timedesc>; rel="next"
...
<
{ [1030 bytes data]
* Connection #0 to host myregistry.azurecr.io left intact
{
  "registry": "myregistry.azurecr.io",
  "imageName": "library/ubuntu",
  "tags": [
    {
      "name": "14.04.1",
      "digest": "sha256:2eb231f768446001a7cf9024ec9724cec23f6768e9ccb4e61499718d34621fbe",
      "createdTime": "2020-03-26T10:09:34.7419234Z",
      "lastUpdateTime": "2020-03-26T10:09:34.7419234Z",
      "signed": false,
      "changeableAttributes": {
        "deleteEnabled": true,
        "writeEnabled": true,
        "readEnabled": true,
        "listEnabled": true
      }
    },
    {
      "name": "xenial-20160217.2",
      "digest": "sha256:c1e33b4b41531cbfc2b6d7c591c9cb228eb6b4cf265b07d9fd1a4b4a0e3d0623",
      "createdTime": "2020-03-26T10:09:32.1301199Z",
      "lastUpdateTime": "2020-03-26T10:09:32.1301199Z",
      "signed": false,
      "changeableAttributes": {
        "deleteEnabled": true,
        "writeEnabled": true,
        "readEnabled": true,
        "listEnabled": true
      }
    },
    {
      "name": "12.04",
      "digest": "sha256:18305429afa14ea462f810146ba44d4363ae76e4c8dfc38288cf73aa07485005",
      "createdTime": "2020-03-26T10:09:31.3938861Z",
      "lastUpdateTime": "2020-03-26T10:09:31.3938861Z",
      "signed": false,
      "changeableAttributes": {
        "deleteEnabled": true,
        "writeEnabled": true,
        "readEnabled": true,
        "listEnabled": true
      }
    }
  ]
}
```

You may refer to the sample script in our doc: https://github.com/Azure/acr/blob/master/docs/AAD-OAuth.md#pagination-1

**Comment by timja** [comment]
@Wwwsylvia 

I'm reusing the existing code:
https://github.com/Azure/acr-cli/blob/e45d021fab521318c2cdd67f59e082667de63414/cmd/acr/purge.go#L278

Let me take a look at the Link header

thanks

**Comment by timja** [comment]
@Wwwsylvia that seems to work, but will need some refactoring here to use pagination correctly. As the code here is just figuring out the last value itself rather than respecting the last parameter

**Comment by Wwwsylvia** [comment]
> @Wwwsylvia
> 
> I'm reusing the existing code:
> https://github.com/Azure/acr-cli/blob/e45d021fab521318c2cdd67f59e082667de63414/cmd/acr/purge.go#L278
> 
> Let me take a look at the Link header
> 
> thanks

Okay... this is interesting... 
Per the Docker v2 spec,
> Compliant client implementations should always use the Link header value when proceeding through results linearly.

So yeah we need to refactor the code.

BTW would you close the support ticket 120072325000556 if no further questions there?

**Comment by timja** [comment]
@Wwwsylvia Is this something that is planned in the short term? As it blocks this being merged I assumed, otherwise we'll have to see if it's something we can do.

Thanks

**Comment by Wwwsylvia** [comment]
@timja Yes let me fix the pagination part these days :)

**Comment by Wwwsylvia** [comment]
Hi @timja, per our latest discussion the refactoring is not planned in a short term. :(
We may have to put this aside for a while, if you mind could you try to include the refactoring in a separate PR? Thanks!

**Comment by Wwwsylvia** [comment]
Hi @timja , the refactor PR is merged. Could you double confirm if this PR has any conflicts?

**Comment by adusumillipraveen** [comment]
> Hi @timja , the refactor PR is merged. Could you double confirm if this PR has any conflicts?

@Wwwsylvia I have merged main after resolving conflicts.

File: cmd/acr/purge.go [hunk 1]
```diff
@@ -272,19 +279,34 @@ func getTagsToDelete(ctx context.Context,
...
 			// as a tag to delete.
 			if lastUpdateTime.Before(timeToCompare) && *(*tag.ChangeableAttributes).DeleteEnabled {
-				tagsToDelete = append(tagsToDelete, tag)
+				tagsEligibleForDeletion = append(tagsEligibleForDeletion, tag)
+			}
+		}
+
+		tagsToDelete := []acr.TagAttributesBase{}
+
+		if keep == 0 {
```
Comment (Wwwsylvia) [inline cmd/acr/purge.go:293]: Wondering if `skippedCount` already reached `keep` here, can we avoid checking check one by one?
```go
newLastTag = getLastTagFromResponse(resultTags)

if keep == 0 || skippedTagsCount == keep {
	tagsToDelete = tagsEligibleForDeletion
	return &tagsToDelete, newLastTag, skippedTagsCount, nil
}
```

File: cmd/acr/purge.go [hunk 2]
```diff
@@ -272,19 +279,34 @@ func getTagsToDelete(ctx context.Context,
...
+			tagsToDelete = tagsEligibleForDeletion
+		} else {
+			for _, tag := range tagsEligibleForDeletion {
+				// Keep at least the configured number of tags
+				if keep-skippedTagsCount == 0 {
```
Comment (Wwwsylvia) [inline cmd/acr/purge.go:298]: For readability, would something like this looks better?
```go
if skippedTagsCount < keep {
	skippedTagsCount++
} else {
	tagsToDelete = append(tagsToDelete, tag)
}
```

File: README.md [hunk 3]
```diff
@@ -159,6 +159,17 @@ acr purge \
     --untagged
 ```
 
+#### Keep flag
+
+To keep at least x number of tags, the `--keep` flag should be set.
```
Comment (Wwwsylvia) [inline README.md:164]: Should we mention the x number is "the latest x number" to make it clear? This applies for other texts.

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (adusumillipraveen) [reply README.md:164]: I am not sure if that gives the exact meaning as its applied on top of ago flag ?

For example, if there are 20 tags in a repo that meet the criteria `--ago 3d` , but when we pass `--keep 10` the logic will still leave 20 tags instead of 10. So, it is making sure **at least 10 tags are kept, but not the latest 10 are kept**

Should we rephrase it to say "at least x number of latest tags"

Comment (Wwwsylvia) [reply README.md:164]: The logic gets tags in timedesc order and skips the fore ones, so it actually keeps the most recent ones of the eligible tags. 
I think the idea is just to let users know which tags to keep. For example, suppose there are eligible tags `a20200102`, `b20200103`, `c20200101`, will `--keep 1` keep `a20200102`(first in alphabetical order) or `b20200103`(latest) or `c20200101`(oldest)? Users may be confused...

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

File: cmd/acr/purge.go [hunk 4]
```diff
@@ -272,19 +279,33 @@ func getTagsToDelete(ctx context.Context,
...
 			}
 			// If a tag did match the regex filter, is older than the specified duration and can be deleted then it is returned
 			// as a tag to delete.
 			if lastUpdateTime.Before(timeToCompare) && *(*tag.ChangeableAttributes).DeleteEnabled {
-				tagsToDelete = append(tagsToDelete, tag)
+				tagsEligibleForDeletion = append(tagsEligibleForDeletion, tag)
 			}
 		}
+
+		tagsToDelete := []acr.TagAttributesBase{}
```
[the remark this answers is not in the document]
Comment (adusumillipraveen) [reply cmd/acr/purge.go]: Nice one

**Comment by adusumillipraveen** [comment]
@Wwwsylvia I have addressed all your comments. Please have a look.

File: README.md [hunk 5]
```diff
@@ -159,6 +159,17 @@ acr purge \
...
 
+#### Keep flag
+
+To keep the latest x number of tags, the `--keep` flag should be set.
+
+```sh
+acr purge \
+    --registry <Registry Name> \
+    --filter <Repository Name>:<Regex filter>
+    --keep 3
```
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply README.md:170]: nit: And there should be `\` mark at the end of `--filter` line.. 
> --filter <Repository Name>:<Regex filter> \

Could you also help to fix other wrong ones in this README, thanks a lot!

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

**Comment by Wwwsylvia** [comment]
@adusumillipraveen  Hi the code looks good, but I still have some concerns on the descriptions... Could you improve those? Thanks!

**Comment by adusumillipraveen** [comment]
> @adusumillipraveen Hi the code looks good, but I still have some concerns on the descriptions... Could you improve those? Thanks!

Done. hopefully you will find them better this time :)

**Review by Wwwsylvia — APPROVED** [review]

**Comment by timja** [comment]
Thanks!
